### PR TITLE
Add support for thrift exceptions in http endpoints

### PR DESCRIPTION
**Pull request by Raynos**
This PR adds support for thrift exceptions in http endpoints.
We now serialize thrift exception structs to JSON with proper
status codes.

I've also added support for generating conversion functions in
the generate workflow that converts client exceptions to server
exceptions.

The conversion code has a TODO where it creates the "default"
struct for the server exception and does not copy over the fields.

r: @uber/zanzibar-team

### CODE REVIEW

**Comment by ChuntaoLu** [comment]
LGTM, seems a rebase is needed.

File: codegen/method.go [hunk 1]
```diff
@@ -304,12 +309,14 @@ func (ms *MethodSpec) setExceptions(
 		}
 
 		if !ms.WantAnnot {
```
Comment (stevededalus) [inline codegen/method.go]: Maybe change WantAnnot to RequireAnnot

File: codegen/method.go [hunk 2]
```diff
@@ -291,6 +293,9 @@ func (ms *MethodSpec) setExceptions(
 		ms.OKStatusCode.Code: true,
 	}
 	ms.Exceptions = make([]ExceptionSpec, len(resultSpec.Exceptions))
+	ms.ExceptionsIndex = make(
```
Comment (stevededalus) [inline codegen/method.go]: May Map or Dict instead of index

File: codegen/method.go [hunk 3]
```diff
@@ -342,6 +349,8 @@ func (ms *MethodSpec) setExceptions(
 				Message: e.Name,
 			},
 		}
+		ms.Exceptions[i] = exception
```
Comment (stevededalus) [inline codegen/method.go]: You might need to sort Exceptions at the end of constructing it to ensure that the ordering in the templates is fixed. I'm not sure what ordering resultSpec.Exceptions will be in.

Comment (Raynos) [inline codegen/method.go]: Seems to not be an issue in practice.

**Pull request merged**